Support GitHub app auth

scottillogical opened the pull request:
# The PR enables using GitHub app authentication

The has the following benefits
- Using github app auth increases the rate limit vs using a personal authentication token.  
- Using github app auth allows you to authenticate via the organization as opposed to an individual. So changes to an individual user account will not affect the pipeline (such as changing organization role from owner to member)

Relevant github issue: https://github.com/telia-oss/github-pr-resource/issues/212
Changes 
* Add new fields to source for github app auth 
* add https://github.com/bdellegrazie/git-credential-github-app to dockerfile

fatmcgav-depop committed:
Add support for authentication as a GitHub App

This commit adds support for authentication as a GitHub App.
This is beneficial in several ways, including:
* Increased rate limits
* Better separation of access
* Finer grained control over access
* Removes the need for a bot or service account

As part of these changes, have added the `github.com/bradleyfalzon/ghinstallation/v2`
module and it's associated dependencies.

Added several new configuration fields:
* `UseGitHubApp` - Boolean flag signalling if user wants to auth as a GitHub App
* `PrivateKeyFile` - Filename for RSA Private Key generated for GitHub App
* `AppID` - GitHub App application numerical identifier
* `InstallationID` - GitHub App installation numerical identifier

Upgrade to golang `1.16`

Add some tests for the `Source` model.

Add support for both `private_key` and `private_key_file` intputs.

This enables either reading the private key from a file, or providing the contents
of the private key.
Expanded testing to cover additional scenarios.

Also rename `AppID` to `ApplicationID`.

add manifest yaml

bump

remove unused travis

use real image tags in dockefile, update go.mod

update go sum

comment xoauth basic

comment xoauth basic

comment xoauth basic

comment xoauth basic

add git credentail helper

fix dockerfile

add credential helper

sds

sds

Commit:
github app support in git clone

remove run e

remove run e

fix gofmt

fix gofmt

fix gofmt

Commit:
fix github org

Commit:
disable access token when using github app

remove logger

Commit:
add github app docs to readme

restore oauth basic

remove private key file reference

typo

punctuation

Update Golang from 1.22.10 to 1.23.7 and its dependencies

bgandon commented on CODEOWNERS:
```diff
@@ -1,3 +0,0 @@
-# Default. Unless we have a more specific match.
```
Please don't remove files unless you have a good reason to do so

scottillogical replied:
restored 52456f6

bgandon commented on Dockerfile:
```diff
@@ -1,33 +1,26 @@
-
-FROM ${golang} AS builder
-
+FROM public.ecr.aws/docker/library/golang:1.22 as builder
```
Please don't change this Dockerfile that is required by the build system in our Concourse pipeline, unless you have good reason to do so.

Instead, you should use a separate Dockerfile to run your tests locally.

scottillogical replied:
Sorry will take a look.  I can see that this has been heavily customized for concourse, but as I hope you can understand that makes it seemingly impossible to build locally

bgandon commented on Dockerfile:
```diff
@@ -1,33 +1,26 @@
 
-RUN go version \
-    && make all
+RUN go version &&  make all
```
No need for such a change

bgandon commented on Dockerfile:
```diff
@@ -1,33 +1,26 @@
 COPY scripts/askpass.sh /usr/local/bin/askpass.sh
+ENV GITHUB_APP_CRED_HELPER_VERSION="v0.3.2"
+ENV BIN_PATH_TARGET=/usr/local/bin
+RUN curl -L https://github.com/bdellegrazie/git-credential-github-app/releases/download/${GITHUB_APP_CRED_HELPER_VERSION}/git-credential-github-app_${GITHUB_APP_CRED_HELPER_VERSION}_Linux_x86_64.tar.gz | tar zxv -C ${BIN_PATH_TARGET}
```
Adding such external component should be justified in the PR description.

Plus, why not use the latest v0.3.3 version?

Which raises the question of how we shall bump this new dependency?

scottillogical replied:
I updated it to the latest version 9f4ae0f7f96a6afe4fd8cb5072b112dba0d7e981

I am open to suggestions for how to keep it up to date.   with say, the git resource, we have to build our own version of the git resource to add support for github app auth, since the regular git resource is agnostic to providers.  but supporting github apps in this resource seems highly desirable

bgandon commented on Dockerfile:
```diff
@@ -1,33 +1,26 @@
+RUN go version &&  make all
 
-
-
```
Please don't modify the spacing that has purposely been adopted for clarity

bgandon commented on Dockerfile:
```diff
@@ -1,33 +1,26 @@
 RUN chmod +x /opt/resource/*
 
-
-
```
Please don't modify the spacing that has purposely been adopted for clarity

bgandon commented on README.md:
```diff
@@ -56,6 +56,11 @@ automated tests.
+| `github_organization`       | No       |  `Vault-tec`                     | Which Github organization your github app is in.
+| `private_key`               | No       |  `-----BEGIN RSA...`             | Private key for your github app.
+| `installation_id`           | No       |  `12356`                         | Installation id for your github app.
+| `application_id`            | No       |  `12356`                         | Application id for your github app.
```
Table rows should be formatted the same as previous row are.

(We typically use the "_format selection_" action on the table, using the `yzhang.markdown-all-in-one` VScode plugin.)

scottillogical replied:
c25444bae4e0ff885aa9a53470c7ee63721e0092

bgandon commented on git.go:
```diff
@@ -75,9 +89,25 @@ func (g *GitClient) Init(branch string) error {
+	} else {
+		if err := g.command("git", "config", "url.https://x-oauth-basic@github.com/.insteadOf", "git@github.com:").Run(); err != nil {
+			return fmt.Errorf("failed to configure github url: %s", err)
+		}
```
Smaller branch of if-then-else should come first: please invert the condition.

scottillogical replied:
0939b215722a1215cae935302b06773e3d4935ee

bgandon commented on github.go:
```diff
@@ -46,23 +47,38 @@ func NewGithubClient(s *Source) (*GithubClient, error) {
+		// Client using oauth2 wrapper
+		client = oauth2.NewClient(ctx, oauth2.StaticTokenSource(
+			&oauth2.Token{AccessToken: s.AccessToken},
+		))
```
Smaller branch of if-then-else should come first: please invert the condition.

scottillogical replied:
37ead8e662b257bb1e9b876a9ae62cc5b696cfe3

bgandon reviewed (changes requested):
Thanks for the contribution!
Here are my overall remarks.

1. Rationale: we'll need to to explain a bit clearer the problem that arises in the first place, and how you solve it with the code you're contributing.

2. Clean code: could you please rebase your work on top of main branch, and clean it up with `git rebase -i` so that you remove unnecessary back-and-forth commits, and remove test-related commits?

3. Failing build: in our Concourse pipeline, the `make all` build step fails with this error:
```
> [builder 5/5] RUN go version &&  make all:
...
#14 37.87 vet: ./models_test.go:42:5: unknown field GithubOrganziation in struct literal of type resource.Source
#14 37.97 make: *** [Makefile:11: test] Error 1
 ```

4. Increased code complexity, yet not covered by unit tests: there are many `if g.UseGithubApp` or `if !g.UseGithubApp` which increases a lot the number of code branches and thus tests cases that should be covered. Un fortunately we don't have sufficient test coverage to support such complexity.

Commit:
revert changes to dockerfile - update github app credential helper

Commit:
invert condition

Commit:
invert condition

Commit:
formatted selection

Commit:
fix github organization

remove unneeded manifest.yaml

scottillogical commented:
> Thanks for the contribution! Here are my overall remarks.
> 
> 1. Rationale: we'll need to to explain a bit clearer the problem that arises in the first place, and how you solve it with the code you're contributing.
> 2. Clean code: could you please rebase your work on top of main branch, and clean it up with `git rebase -i` so that you remove unnecessary back-and-forth commits, and remove test-related commits?
> 3. Failing build: in our Concourse pipeline, the `make all` build step fails with this error:
> 
> ```
> > [builder 5/5] RUN go version &&  make all:
> ...
> #14 37.87 vet: ./models_test.go:42:5: unknown field GithubOrganziation in struct literal of type resource.Source
> #14 37.97 make: *** [Makefile:11: test] Error 1
> ```
> 
> 4. Increased code complexity, yet not covered by unit tests: there are many `if g.UseGithubApp` or `if !g.UseGithubApp` which increases a lot the number of code branches and thus tests cases that should be covered. Un fortunately we don't have sufficient test coverage to support such complexity.

1. Added "why" to the PR description
2. Rebased my commits
3. Fixed build failures
4. I added some test coverage
